table_helper.cc

```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
 }
 
 future<> table_helper::cache_table_info(service::query_state& qs) {
-    if (_prepared_stmt) {
-        return now();
-    } else {
+    if (!_prepared_stmt) {
         // if prepared statement has been invalidated - drop cached pointers
         _insert_stmt = nullptr;
+    } else if (!_is_fallback_stmt) {
```

Comment:
I have a hard time parsing this if-else. I see following cases:
!_prepared_stmt -> we prepare
_prepared_stmt && !_is_fallback_stmt -> return because we have already prepared the statement
_prepared_stmt && _is_fallback_stmt -> we have fallback statement prepared. Why do we want to re-prepare in this case?

Reply:
We've used the fallback statement because we couldn't use the regular statement (the table isn't updated yet). We want to start using the regular statement when the table gets updated, so we'll try to prepare it even if we have a successfully prepared fallback statement. If we fail again, we check for _prepared_stmt && _is_fallback_stmt to avoid re-preparing the fallback statement.

Reply:
Thanks for the explanation. Probably worth leaving a comment like that in the code.

Reply:
And what if tomorrow we want to add one more column or N more? What will that be:
```
if (statement) ... else if (fallback1_statement)...else if (fallback2_statement)....if (fallbackN_statement)...
```

This obviously doesn't scale.
This fallback game leads to a very not-pretty spaghetti-like code hard to read and maintain.

If we believe that we want to add this new column let's simply add it and whoever haven't updated their schema will not be able to use Tracing until he/she does (updates the schema).

This is exactly how it works today and the only complain there is - there no enough documentation and some packaging didn't include the `fix_system_distributed_tables.py` script: https://github.com/scylladb/scylla/issues/4601

I don't see a reason to change an approach in general and this "fallback" proposal doesn't look like a good alternative to me either.

Reply:
I'm ok with dropping the fallback and just relying on the script. @avikivity what about you? You proposed the fallback in https://github.com/scylladb/scylla/pull/6827#discussion_r454158997 so the question is whether you're ok with not doing it.

Reply:
First, there will always be a window of time during cluster upgrade where we need the fallback.

Second, we don't enforce running the script, so we can be left with a cluster that doesn't support tracing.

Reply:
> First, there will always be a window of time during cluster upgrade where we need the fallback.
We will need the fallback only if tracing is enabled or any table, right?
I guess forcing users to ensure no queries with tracing are running during rolling upgrade is quite a burden and will back fire sooner or later.

> 
> Second, we don't enforce running the script, so we can be left with a cluster that doesn't support tracing.

@vladzcloudius any counter arguments or should we just merge as it is? I like the simplest solution possible but if it's not sufficient maybe the complexity is worth it. How are the problems mentioned by @avikivity handled right now by the clients?

Reply:
> > First, there will always be a window of time during cluster upgrade where we need the fallback.
> 
> We will need the fallback only if tracing is enabled on any table, right?

Yes.

> I guess forcing users to ensure no queries with tracing are running during rolling upgrade is quite a burden and will back fire sooner or later.

Application development and cluster operation can be completely separated. The cluster operator can even be a machine (the kubernetes operator). Even the application may decide to trace on its own (okay, that's not very likely). But it remains that forcing the application developer to synchronize with cluster operations is not good.

Reply:
> > > First, there will always be a window of time during cluster upgrade where we need the fallback.
> > 
> > 
> > We will need the fallback only if tracing is enabled on any table, right?
> 
> Yes.
> 
> > I guess forcing users to ensure no queries with tracing are running during rolling upgrade is quite a burden and will back fire sooner or later.
> 
> Application development and cluster operation can be completely separated. The cluster operator can even be a machine (the kubernetes operator). Even the application may decide to trace on its own (okay, that's not very likely). But it remains that forcing the application developer to synchronize with cluster operations is not good.

It's hard to argue with this point. We should be doing all we can to automate the operations and not impose manual procedures and checklists. If I were to choose, I would go with a solution that does not involve user actions.

Reply:
@avikivity @haaawk The chance for such a scenario (devops are executing an upgrade while somebody debugs something at the same time using tracing is close to 0).
In a case of such an unlikely event during the window mentioned by Avi before the same engineer who debugs will probably have to wait till the end of the upgrade. - He'll probably have to wait for that due to multiple other reasons. Therefore there shouldn't be any drama if he forgets to stop traces and some of them fail.

But look from the other side - if we merge this fallback thingy we will have to maintain it.
What if we decide to add a few more columns in the future? How many fallback branches do we want to maintain?

I think it's safe enough not to worry about a few failed traces during the upgrade instead of merging this code which has a good potential to become a spaghetti.

Reply:
Personally, I'm not in favor of the spaghetti code argument. If in the future we want/need to add more columns then we deal with it then. The code can always be refactored. The decision about having fallbacks can always be reverted unless we make it public (which we don't plan to do). Let's not over-engineer. At the moment it's a single `if` and that's what we're talking about. Hypothesizing is not a graceful base for arguments because it also powers argument like "what if in the future we have a potential client that wants to pay us 1B USD but requires tracing to always work". Also, we can remove the fallback code in the next release after the one that introduces the new column, no?

So far we have the following arguments:
For fallback:
1. Not having it opens up a can of worms related to human errors like DevOps team forgetting to run the script, cluster running for days, weeks, months, then having an issue which someone (potentially @vladzcloudius or someone less experienced) tries to debug with tracing but it fails. In such potentially stressful situation it is definitely something desirable. Not only will it slow down the process of solving the issue but could also hurt customer confidence in us.
2. It may be not ok for tracing to not work during the rolling upgrade.

Against fallback:
1. Code is simpler without callback
2. It may be ok for a query with tracing to fail (is it really?) for a short (is it really short) period of time the cluster is in rolling upgrade

I'm not listing the 'in the future' argument because I don't think it's a problem. The code can always be adjusted to future needs.
The existence of a proper solution based on transactions is kind of neutral because it means we will be able to remove the fallback code once we have transactions but no one really knows when that will be. In the meantime, we will have to deal with customer clusters for quite some time.

My suggestion is to merge the fallback code with a comment to remove it after one release (not the minor release like 4.1.1->4.1.2). That gives us most of the benefits. Tracing will work during a rolling upgrade and the code will still be simpler in the long run. The only thing we won't get is the protection in case of the situation the DevOps team forgets to run the script. In such case tracing will work until the next rolling upgrade (but not minor so until 4.1->4.2, not until 4.1.1->4.1.2) if they forget to run the script again. Not ideal but at least reduces the chances by giving the DevOps team few events (minor version upgrades and one non-minor upgrade) to think about running this script. @avikivity @vladzcloudius what do you think?

Reply:
We could also check with the new version that the tracing table has the right schema and refuse to boot if not.

Reply:
> @avikivity @haaawk The chance for such a scenario (devops are executing an upgrade while somebody debugs something at the same time using tracing is close to 0).
> In a case of such an unlikely event during the window mentioned by Avi above the engineer who debugs will probably have to wait till the end of the upgrade. - More than that, he'll probably have to wait due to multiple other reasons beyond tracing. Therefore there shouldn't be any drama if he forgets to stop traces and some of them fail.
> 
> But look from the other side - if we merge this fallback thingy we will have to maintain it.
> What if we decide to add a few more columns in the future? How many fallback branches do we want to maintain?
> 

As many as needed. We've added plenty of cluster features to make upgrades painless.

> I think it's safe enough not to worry about a few failed traces during the upgrade instead of merging this code which has a good potential to become spaghetti.

How did you arrive at "a few"? It could be a large number since cluster upgrade is not bounded in time, and there is nothing that will add the missing column automatically.

> 
> A proper way to address this is to make sure that a schema change is a transaction (on top of Paxos or Raft) and then trigger a schema change during the first boot after the upgrade.

This is not something we can do now.

Reply:
> Personally, I'm not in favor of the spaghetti code argument. If in the future we want/need to add more columns then we deal with it then. The code can always be refactored. The decision about having fallbacks can always be reverted unless we make it public (which we don't plan to do). Let's not over-engineer. At the moment it's a single `if` and that's what we're talking about. Hypothesizing is not a graceful base for arguments because it also powers argument like "what if in the future we have a potential client that wants to pay us 1B USD but requires tracing to always work". Also, we can remove the fallback code in the next release after the one that introduces the new column, no?
> 
> So far we have the following arguments:
> For fallback:
> 
> 1. Not having it opens up a can of worms related to human errors like DevOps team forgetting to run the script, cluster running for days, weeks, months, then having an issue which someone (potentially @vladzcloudius or someone less experienced) tries to debug with tracing but it fails. In such potentially stressful situation it is definitely something not desirable. Not only will it slow down the process of solving the issue but could also hurt customer confidence in us.
> 2. It may be not ok for tracing to not work during the rolling upgrade.
> 
> Against fallback:
> 
> 1. Code is simpler without callback

IMO the extra cost in complexity is minor. An "if" statement is not something beyond our capability.

> 2. It may be ok for a query with tracing to fail (is it really?) for a short (is it really short?) period of time the cluster is in rolling upgrade

First, upgrade is not a short time.

Second, it is a stressful period, and having an extra source of failures is making it worse.

Third, we don't yet have an automatic way of adding that column.

> 
> I'm not listing the 'in the future' argument because I don't think it's a problem. The code can always be adjusted to future needs.
> The existence of a proper solution based on transactions is kind of neutral because it means we will be able to remove the fallback code once we have transactions but no one really knows when that will be. In the meantime, we will have to deal with customer clusters for quite some time.
> 
> **My suggestion is to merge the fallback code with a comment to remove it after one release** (not the minor release like 4.1.1->4.1.2). That gives us most of the benefits. Tracing will work during a rolling upgrade and the code will still be simpler in the long run. The only thing we won't get is the protection in case of the situation the DevOps team forgets to run the script. In such case tracing will work until the next rolling upgrade (but not minor so until 4.1->4.2, not until 4.1.1->4.1.2) if they forget to run the script again. Not ideal but at least reduces the chances by giving the DevOps team few events (minor version upgrades and one non-minor upgrade) to think about running this script. @avikivity @vladzcloudius what do you think?

That code should live on for a year, maybe two. Like all of the compatibility hacks. See 253a7640e3c770d939ae0024ee072a76d8d9c6cc which cleaned up two year old cluster features. This is no different.

I am confident in our ability to maintain that if statement, and perhaps even add another branch in the event that we add another column in the next year or two.

Reply:
Considering there has been no discussion in the last couple of days, is it safe to assume that we agree to accept the current version?
Personally, I don't have a strong preference of whether we have to add the fallback, but if I were to choose, I think I'm more convinced by @avikivity's and @haaawk's arguments. If we don't add the additional statement, we will have to rely entirely on the script, and it looks like there were no changes in the columns in the last 2 years (at least there were no changes in the tables in the script), so users didn't need to run it and they might not even be aware that it exists (especially because there is no documentation about it).
All in all, it feels like we're choosing between a minor inconvenience for us and a minor inconvenience for the users, and in such a case, I'm pretty sure we'd prefer the former.

Reply:
@wmitros 
All it takes is to run a script before you restart the service after the upgrade and that will make no extra hacks needed. Everything will continue to work before and after the upgrade.

Our upgrade documentation is per-upgrade version, so the next one that relates to the version with this new column should have the corresponding bullet at the corresponding place. 

> How did you arrive at "a few"? It could be a large number since cluster upgrade is not bounded in time, and there is nothing that will add the missing column automatically.

@avikivity Rephrasing the Mrs. May popular saying: "It's highly unlikely that the number of these requests is going to be greater than 0." 
However in order to cover such an unlikely event one should follow the procedure that we should have in our upgrade notes as I described above.

If the script is executed before the first scylla start after the upgrade then it will be exactly 0 without and fallbacks.

However I see that you have merged it already, so - now we have (more) things to cleanup following  253a764.